File: CodingSamples/DotNet/Language/InterfaceTest2/DemoApp/Program1.cs
using Taxation;

class Program
{
    //nested class - can only refer to static members
    //of its outer class
    class Auditor
    {
        public Auditor()
        {
            Console.WriteLine("Auditing begins...");
        }

        public void Audit(string id, ITaxPayer emp)
        {
            if(id.Length < 4)
                throw new ArgumentException("Invalid ID");
            decimal payment = emp.IncomeTax() + 250;
            Console.WriteLine("Total Tax Payment: {0:0.00}", payment);
        }

        public void Dispose()
        {
            Console.WriteLine("Auditing ends.");
        }
    }

    static void Action(string name, int count)
    {
        ITaxPayer t = name == "jack" ? new Supervisor { Span = count } : new Worker { Jobs = count };
        Auditor a = new Auditor();
        try
        {
            a.Audit(name, t);
        }
        finally
        {
            a.Dispose();
        }
        
    }

    static void Main(string[] args)
    {
        try
        {
            string m = args[0].ToLower();
            int n = int.Parse(args[1]);
            Action(m, n);
        }
        catch(Exception ex)
        {
            Console.WriteLine("Action Failed: {0}", ex.Message);
        }
    }
}

File: CodingSamples/Java/Language/InterfaceTest2/Program1.java
import taxation.*;

class Program {

	//inner nested member class - a class defined inside another class
	//with static modifier so that it can only refer to static members
	//of its outer class
	static class Auditor {

		public Auditor() {
			System.out.println("Auditing begins...");
		}

		public void audit(String id, TaxPayer emp) {
			if(id.length() < 4)
				throw new IllegalArgumentException("Invalid ID");
			double payment = emp.incomeTax() + 250;
			System.out.printf("Total Tax Payment: %.2f%n", payment);
		}

		public void close() {
			System.out.println("Auditing ends.");
		}

	}

	private static void action(String name, int count) {
		TaxPayer t = name.equals("jack") ? new Supervisor(count) : new Worker(count);
		Auditor a = new Auditor();
		try{
			a.audit(name, t);
		}finally{
			a.close();
		}
	}

	public static void main(String[] args) {
		try{
			String name = args[0].toLowerCase();
			int count = Integer.parseInt(args[1]);
			action(name, count);
		}catch(Exception e){
			System.out.printf("Action failed: %s%n", e.getMessage());
		}
	}
}

